server/controllers/nlp.js

```diff
@@ -148,6 +199,57 @@ function findDataTransformationFunction(state, callback, errorCallback) {
         }
         errorCallback(errorMessage)
     } else {
-        callback({command: bestMatchedCommand, parameters: columnsArray})
+        callback({command: bestMatchedCommand, parameters: {columns: columnsArray, filters: filterArray}})
+    }
+}
+
+
+/**
+ * takes the list of (simplified) tokenHolders from the state, combines them by predefined rules to filters and groupings and puts the new structure back into the state
+ * @param state
+ */
+function combineComplexTokens(state) {
+    // hide unclassified tokens
+    state.tokenHolders = state.tokenHolders.filter(tokenHolder => tokenHolder.label != null || tokenHolder.labelType === Classifier.staticWords.value);
+
+    // include ColumnValues/Values into FilterSelectors
+    for (let i = 0; i < state.tokenHolders.length - 1; i++) {
+        if ((state.tokenHolders[i].labelType === Classifier.staticWords.filterSelector
+            && (state.tokenHolders[i + 1].labelType === Classifier.staticWords.value || state.tokenHolders[i + 1].labelType === Classifier.staticWords.columnValue))
+            || (state.tokenHolders[i].labelType === Classifier.staticWords.genericSelector && state.tokenHolders[i + 1].labelType === Classifier.staticWords.columnValue)) {
+            state.tokenHolders[i].filterValue = state.tokenHolders[i + 1];
+            state.tokenHolders.splice(i + 1, 1);
+
+            if (state.tokenHolders[i].filterValue.column === null && i > 0 && state.tokenHolders[i - 1].labelType === Classifier.staticWords.column) {
+                state.tokenHolders[i].filterValue.column = state.tokenHolders[i - 1].label
+            }
+            // TODO verify that (assigning the same column as the previous filter)
+            if (state.tokenHolders[i].filterValue.column === null && i > 0 && (state.tokenHolders[i - 1].labelType === Classifier.staticWords.filterSelector || state.tokenHolders[i - 1].labelType === Classifier.staticWords.genericSelector)) {
```

Comment:
Maybe include comments in long chained conditions, otherwise it is hard to understand them

Reply:
done